python-sdk/nuscenes/predict/helper.py

```diff
@@ -0,0 +1,181 @@
+
+from typing import Dict, Tuple, Any, List
+from pyquaternion import Quaternion
+from nuscenes import NuScenes
+from nuscenes.utils.geometry_utils import transform_matrix
+from nuscenes.eval.common.utils import quaternion_yaw
+import numpy as np
+
+MICROSECONDS_PER_SECOND = 1e6
+BUFFER = 0.15 # seconds
+
+def angle_of_rotation(yaw: float) -> float:
+    return (np.pi / 2) + np.sign(-yaw) * np.abs(yaw)
+
+def convert_global_coords_to_local(coordinates: np.ndarray,
+                                   translation: Tuple[float, float, float],
+                                   rotation: Tuple[float, float, float, float]) -> np.ndarray:
+    """Converts global coordinates to coordinates in the frame given by the rotation quaternion and
+    centered at the translation vector.
+    :param coordinates: x,y locations. array of shape [n_steps, 2]
+    :param translation: Tuple of (x, y, z) location that is the center of the new frame
+    :param rotation: Tuple representation of quaternion of new frame."""
+
+    yaw = angle_of_rotation(quaternion_yaw(Quaternion(rotation)))
+
+    transform = np.array([[np.cos(yaw), -np.sin(yaw)], [np.sin(yaw), np.cos(yaw)]])
+
+    coords = (coordinates - np.atleast_2d(np.array(translation)[:2])).T
+
+    return np.dot(transform, coords).T[:, :2]
+
+class PredictHelper:
+    """Wrapper class around NuScenes to help retrieve data for the prediction task."""
+
+    def __init__(self, nusc: NuScenes):
+        """Inits PredictHelper
+        :param nusc: Instance of NuScenes class
+        """
+        self.data = nusc
+        self.inst_sample_to_ann = self._map_sample_and_instance_to_annotation()
+
+    def _map_sample_and_instance_to_annotation(self) -> Dict[Tuple[str, str], str]:
+        """Creates mapping to look up an annotation given a sample and instance in constant time."""
+        mapping = {}
+
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str):
+        """"""
+        return self.data.get('sample', sample)['timestamp']
+
+    def _absolute_time_diff(self, time1: int, time2: int) -> int:
+        """Helper to compute how much time has elapsed in _iterate method."""
+        return abs(time1 - time2) / MICROSECONDS_PER_SECOND
+
+    def _iterate(self, starting_annotation: Dict[str, Any], seconds: float, direction: str) -> List[Dict[str, Any]]:
+        """Iterates forwards or backwards in time through the annotations for a given amount of seconds."""
+
+        seconds_with_buffer = seconds + BUFFER
```

Comment:
Might want to do a sanity check that `seconds` is non-negative.

Reply:
Done!